scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Policy.scala

```diff
@@ -99,13 +115,67 @@ object Policy {
         )
     }
 
+    override def unexpired(ft: FormatToken): Policy =
+      p1.unexpired(ft) & p2.unexpired(ft)
+
     override def filter(pred: Clause => Boolean): Policy =
       p1.filter(pred) & p2.filter(pred)
 
     override def exists(pred: Clause => Boolean): Boolean =
       p1.exists(pred) || p2.exists(pred)
 
+    override def noDequeue: Boolean =
+      p1.noDequeue || p2.noDequeue
+
     override def toString: String = s"($p1 & $p2)"
   }
 
+  object Proxy {
+    def apply(
+        policy: Policy
+    )(factory: Policy => Pf)(implicit line: sourcecode.Line): Policy =
+      if (policy.isEmpty) NoPolicy
+      else new Proxy(policy, factory)
+  }
+
+  private class Proxy(
+      policy: Policy,
+      factory: Policy => Policy.Pf
+  )(implicit line: sourcecode.Line)
+      extends Policy {
+    override val f: Pf = factory(policy)
+
+    override def exists(pred: Clause => Boolean): Boolean = policy.exists(pred)
+
+    override def filter(pred: Clause => Boolean): Policy =
+      Proxy.apply(policy.filter(pred))(factory)
+
+    override def unexpired(ft: FormatToken): Policy =
+      Proxy.apply(policy.unexpired(ft))(factory)
+
+    override def noDequeue: Boolean = policy.noDequeue
+
+  }
+
+  sealed abstract class End {
```

Comment:
nit

```suggestion
  sealed trait End {
```

Reply:
why trait? i was under impression that abstract classes are a little more efficient than traits in calling methods.

Reply:
Wow, i don't know nothing about it. Can you please share an article or benchmarks about it?

I prefer trait here because usually the only reason to use abstract class in sealed hierarchies is passing parameters to constructor. And here there is no args

Reply:
https://www.scala-lang.org/blog/2016/07/08/trait-method-performance.html
https://shipilev.net/blog/2015/black-magic-method-dispatch/

will change to trait for now.